File: go-code/leetcode/problem_2.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	root := &ListNode{}
	next := root
	v := 0
	for l1 != nil && l2 != nil {
		v += l1.Val + l2.Val
		next.Next = &ListNode{
			Val: v % 10,
		}
		v /= 10
		next = next.Next
		l1 = l1.Next
		l2 = l2.Next
	}
	for l1 != nil {
		v += l1.Val
		next.Next = &ListNode{
			Val: v % 10,
		}
		v /= 10
		next = next.Next
		l1 = l1.Next
	}
	for l2 != nil {
		v += l2.Val
		next.Next = &ListNode{
			Val: v % 10,
		}
		v /= 10
		next = next.Next
		l2 = l2.Next
	}
	if v > 0 {
		next.Next = &ListNode{
			Val: v,
		}
	}
	return root.Next
}

File: cpp-code/leetcode/problem_2.cpp
struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* root = new ListNode();
        ListNode* next = root;
        int v = 0;
        while (l1 != nullptr && l2 != nullptr) {
            v += l1->val + l2->val;
            next->next = new ListNode(v % 10);
            v /= 10;
            next = next->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        while (l1 != nullptr) {
            v += l1->val;
            next->next = new ListNode(v % 10);
            v /= 10;
            next = next->next;
            l1 = l1->next;
        }
        while (l2 != nullptr) {
            v += l2->val;
            next->next = new ListNode(v % 10);
            v /= 10;
            next = next->next;
            l2 = l2->next;
        }
        if (v > 0) {
            next->next = new ListNode(v);
        }
        return root->next;
    }
};